File: python/mirage/server/version/state_diff.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.server.version.api import read_version, resolve_ref, version_diff
from mirage.server.version.state_tree import to_state
from mirage.server.version.store import VersionStore
from mirage.types import SessionKey, StateKey


def _dict_delta(before: dict[str, Any], after: dict[str,
                                                    Any]) -> dict[str, Any]:
    added = {k: after[k] for k in after.keys() - before.keys()}
    deleted = {k: before[k] for k in before.keys() - after.keys()}
    modified = {
        k: {
            "from": before[k],
            "to": after[k]
        }
        for k in before.keys() & after.keys() if before[k] != after[k]
    }
    return {"added": added, "deleted": deleted, "modified": modified}


def _is_empty(delta: dict[str, Any]) -> bool:
    return not (delta["added"] or delta["deleted"] or delta["modified"])


def _session_delta(before: dict[str, Any], after: dict[str,
                                                       Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    env = _dict_delta(
        before.get(SessionKey.ENV) or {},
        after.get(SessionKey.ENV) or {})
    if not _is_empty(env):
        out["env"] = env
    grants = _dict_delta(
        before.get("mount_modes") or {},
        after.get("mount_modes") or {})
    if not _is_empty(grants):
        out["mount_modes"] = grants
    if before.get(SessionKey.CWD) != after.get(SessionKey.CWD):
        out["cwd"] = {
            "from": before.get(SessionKey.CWD),
            "to": after.get(SessionKey.CWD),
        }
    return out


def _sessions_diff(before: list[dict[str, Any]],
                   after: list[dict[str, Any]]) -> dict[str, Any]:
    a = {s[SessionKey.SESSION_ID]: s for s in before}
    b = {s[SessionKey.SESSION_ID]: s for s in after}
    modified = {}
    for sid in a.keys() & b.keys():
        delta = _session_delta(a[sid], b[sid])
        if delta:
            modified[sid] = delta
    return {
        "added": sorted(b.keys() - a.keys()),
        "deleted": sorted(a.keys() - b.keys()),
        "modified": modified,
    }


def _commands_between(history_a: list[dict[str, Any]],
                      history_b: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """The command trail from state A to state B: events in B's history
    that A does not already contain (identity by full event content, so
    replays and clears stay honest)."""
    seen = {tuple(sorted(e.items())) for e in history_a}
    return [e for e in history_b if tuple(sorted(e.items())) not in seen]


async def state_diff(store: VersionStore, ref_a, ref_b) -> dict[str, Any]:
    """The structured difference, category by category, between two versions.

    The judge's evidence surface: files (content), sessions (env
    references, mount grants with from/to, cwd), namespace nodes
    (symlinks and overlays), and the commands that ran between the two
    states. Wire shape is plain snake_case JSON, agent-legible by
    design.
    """
    version_a = await resolve_ref(store, ref_a)
    version_b = await resolve_ref(store, ref_b)
    entries_a, meta_a = await read_version(store, version_a)
    entries_b, meta_b = await read_version(store, version_b)
    state_a = to_state(entries_a, meta_a)
    state_b = to_state(entries_b, meta_b)

    sessions = _sessions_diff(
        state_a.get(StateKey.SESSIONS) or [],
        state_b.get(StateKey.SESSIONS) or [])
    namespace = _dict_delta(
        state_a.get(StateKey.NODES) or {},
        state_b.get(StateKey.NODES) or {})
    commands = _commands_between(
        state_a.get(StateKey.HISTORY) or [],
        state_b.get(StateKey.HISTORY) or [])
    return {
        "files": await version_diff(store, version_a, version_b),
        "sessions": sessions,
        "namespace": namespace,
        "commands": commands,
    }

File: typescript/packages/server/src/version/stateDiff.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { readVersion, resolveRef, versionDiff } from './api.ts'
import { toState, type AnyDict } from './stateTree.ts'
import type { VersionStore } from './store.ts'
import { compareCodePoints } from '@struktoai/mirage-core/utils/sort'

interface DictDelta {
  added: AnyDict
  deleted: AnyDict
  modified: Record<string, { from: unknown; to: unknown }>
}

interface SessionsDiff {
  added: string[]
  deleted: string[]
  modified: Record<string, AnyDict>
}

function dictDelta(before: AnyDict, after: AnyDict): DictDelta {
  const added: AnyDict = {}
  const deleted: AnyDict = {}
  const modified: Record<string, { from: unknown; to: unknown }> = {}
  for (const k of Object.keys(after)) {
    if (!(k in before)) added[k] = after[k]
  }
  for (const k of Object.keys(before)) {
    if (!(k in after)) deleted[k] = before[k]
    else if (JSON.stringify(before[k]) !== JSON.stringify(after[k])) {
      modified[k] = { from: before[k], to: after[k] }
    }
  }
  return { added, deleted, modified }
}

function isEmpty(delta: DictDelta): boolean {
  return (
    Object.keys(delta.added).length === 0 &&
    Object.keys(delta.deleted).length === 0 &&
    Object.keys(delta.modified).length === 0
  )
}

// The session's own id is the key the delta is filed under, and
// `generation` is a counter that bumps on every line the session runs, so
// reporting it would mark every session modified after any activity at all.
// Everything else the session serializes is state a reader may see move.
const SESSION_DIFF_SKIP = new Set(['session_id', 'generation'])

function isPlainDict(value: unknown): boolean {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

// Every session field that moved between the two payloads. Naming three
// fields (env, mount_modes, cwd) meant the diff went quiet on every other
// one, so a version whose only change was an access rule or a standing
// answer read as unmodified right up to the checkout that applied it.
// Walking the union keeps a field added later in view without a second
// list to remember it.
function sessionDelta(before: AnyDict, after: AnyDict): AnyDict {
  const out: AnyDict = {}
  const keys = [...new Set([...Object.keys(before), ...Object.keys(after)])]
  for (const key of keys.sort(compareCodePoints)) {
    if (SESSION_DIFF_SKIP.has(key)) continue
    const was = before[key]
    const now = after[key]
    if (JSON.stringify(was) === JSON.stringify(now)) continue
    if (isPlainDict(was) || isPlainDict(now)) {
      const delta = dictDelta(
        (was as AnyDict | undefined) ?? {},
        (now as AnyDict | undefined) ?? {},
      )
      if (!isEmpty(delta)) out[key] = delta
      continue
    }
    // A field one table lacks reads as null, as the python side renders it.
    out[key] = { from: was ?? null, to: now ?? null }
  }
  return out
}

function sessionsDiff(before: AnyDict[], after: AnyDict[]): SessionsDiff {
  const a = new Map(before.map((s) => [s.session_id as string, s]))
  const b = new Map(after.map((s) => [s.session_id as string, s]))
  const modified: Record<string, AnyDict> = {}
  for (const [sid, sb] of a) {
    const sa = b.get(sid)
    if (sa === undefined) continue
    const delta = sessionDelta(sb, sa)
    if (Object.keys(delta).length > 0) modified[sid] = delta
  }
  return {
    added: [...b.keys()].filter((sid) => !a.has(sid)).sort(compareCodePoints),
    deleted: [...a.keys()].filter((sid) => !b.has(sid)).sort(compareCodePoints),
    modified,
  }
}

// The command trail from state A to state B: events in B's history that
// A does not already contain (identity by full event content, so
// replays and clears stay honest).
function commandsBetween(historyA: AnyDict[], historyB: AnyDict[]): AnyDict[] {
  const key = (e: AnyDict): string =>
    JSON.stringify(Object.entries(e).sort(([x], [y]) => compareCodePoints(x, y)))
  const seen = new Set(historyA.map(key))
  return historyB.filter((e) => !seen.has(key(e)))
}

/**
 * The structured difference, category by category, between two versions.
 *
 * The judge's evidence surface: files (content), sessions (env
 * references, mount grants with from/to, cwd), namespace nodes
 * (symlinks and overlays), and the commands that ran between the two
 * states. Wire shape is plain snake_case JSON with git's
 * added/modified/deleted vocabulary at every level. Mirrors the
 * Python state_diff.
 */
export async function stateDiff(store: VersionStore, refA: string, refB: string): Promise<AnyDict> {
  const versionA = await resolveRef(store, refA)
  const versionB = await resolveRef(store, refB)
  const a = await readVersion(store, versionA)
  const b = await readVersion(store, versionB)
  const stateA = toState(a.entries, a.meta) as unknown as AnyDict
  const stateB = toState(b.entries, b.meta) as unknown as AnyDict

  const sessions = sessionsDiff(
    (stateA.sessions as AnyDict[] | undefined) ?? [],
    (stateB.sessions as AnyDict[] | undefined) ?? [],
  )
  return {
    files: await versionDiff(store, versionA, versionB),
    sessions,
    namespace: dictDelta(
      (stateA.nodes as AnyDict | undefined) ?? {},
      (stateB.nodes as AnyDict | undefined) ?? {},
    ),
    commands: commandsBetween(
      (stateA.history as AnyDict[] | undefined) ?? [],
      (stateB.history as AnyDict[] | undefined) ?? [],
    ),
  }
}
